### ISSUE TITLE
[enhancement] For tags and numerics, return index value if no document is saved

### ISSUE THREAD

**Issue by tw-bert** [RedisLabsModules/RediSearch#292]
For tags and numerics, please return the in-index value if no document is linked to the index row.

Clarified:  
RediSearch `FT.ADD` has the `NOSAVE` option.  
We hoped to use it on some huge datasets, which do need indexing, but because of the size: every byte counts. So a single numeric value should be stored only once (in the FT index), and not in an extra redis Hash key, nor in a FT Payload. Even the document name (which **is** stored and **can** be retrieved) is no option for us here (it violates our row unicity requirements).
 
While the index with `NOSAVE` works fine, the values are not returned, when there is no document (and there is only an index).

Would it be possible to return the index field content if the doc is missing, for `NUMERIC` and `TAG` types? Sidenote: we don't mention `TEXT` type only because **we** don't see common need for collated and/or tokenized and/or stemmed text returned, but it wouldn't hurt either. 

Possible api alternative: add `SKIPDOC` as an option to `FT.SEARCH`.

**Comment by tw-bert** [comment]
Addendum: Where I say 'index', also `NOINDEX SORTABLE` should be considered (I know, could be tricky).

**Comment by dvirsky** [comment]
I'm actually considering it for the aggregation engine, it will speed things up and save memory. The alternative right now is to use sortable for numeric, and for tags there isn't an alternative.

But I'm definitely considering it. It's not so easy (at least for tags)

**Comment by dvirsky** [comment]
Oh, another thing to note - for tags, multiple values may exist.

**Comment by tw-bert** [comment]
For multiple tags, it might be worth considering returning only the first tag 'hit'. This might make sense to guarantee maximum performance.

**Comment by tw-bert** [comment]
> use sortable for numeric

That currently would only work for one field (which in our scenarios never happens, we create custom sort-fields to work around the only-one-sortby-field limitation)

**Comment by dvirsky** [comment]
So you might actually have good use for the aggregation API after all. It
can be used to sort by multiple fields and do mathematic transformations on
them before or after sorting.

**Comment by tw-bert** [comment]
@dvirsky Wow. I will take a good look at it in a week or so (taking a short break). Excellent news.

**Comment by dvirsky** [comment]
yeah it's pretty powerful. You can sort by the log of the ratio of two values if you want :)

**Comment by tw-bert** [comment]
Hahaha, I'll try to get it to sort recursively into time-space infinity (whenever it all goes too fast for us). But seriously : can I sort on multiple fields **and** get multiple WITHSORTKEYS? I guess we'd need that when using NOSAVE?

**Comment by dvirsky** [comment]
I'm going to start adding some docs tomorrow, so you might want to wait a
day or two.

**Comment by tw-bert** [comment]
Take your time, you and your team always work at such an amazing speed.

**Comment by dvirsky** [comment]
@tw-bert enjoy... https://gist.github.com/dvirsky/fa1cf6e65f3269bfb847fb7857151b23

**Comment by tw-bert** [comment]
@dvirsky Very impressive, good stuff

**Comment by dvirsky** [comment]
I'm closing this as it can now be done with the aggregation API

**Comment by dvirsky** [comment]
Oh, for numeric it doesn't work still. For text and tags there's the `matched_terms()` projection that returns the matched terms even if they weren't saved. e.g.:

```
FT.AGGREGATE myIndex "@tag:{foo | bar | baz }"
   APPLY matched_terms() AS term
   GROUPBY 1 @term
      REDUCE count 0
```

**Comment by tw-bert** [comment]
Great info, `matched_terms()` it is.

We have a usecase where we'd like to index about 10,000,000,000 floats in RediSearch (tagged and indexed, about 5000 tags and about 2,000,000 per group within a tag). 

For this usecase, it would be great to have support for numeric + `NOSAVE`.
Can I re-open this issue (enhancement request), or would you like me to submit a new one?

**Comment by dvirsky** [comment]
Let me see if I can just quickly make it apply to numerics as well.
Shouldn't be much work.

**Comment by dvirsky** [comment]
Hmmm... well, it's not that easy, and matched_terms does not cover everything you need. 

The thing is, it doesn't tell you which field matched the terms if you're searching on multiple fields. It just gives you the terms for each record. So while adding the numeric values to this array is trivial, if you for instance search for `@tag:{foo | bar} @numeric[1 500]`, the resulting array will just be `["foo", 135]`, which is not very practical.

I think the best way to solve this would be to add something like `matched_value(@field)` which will be more specific than matched_terms and can only be used for non text matches. It's not a huge change but something beyond the scope of what I'm going to finish before leaving Redis Labs in a few days. 

@mnunberg ping me if you want to discuss how to best implement this in the future.

**Comment by tw-bert** [comment]
@dvirsky Thanks for looking into it. Another option we considered in Feb 2018, was: Just let `FT.SEARCH` return the index value if it can't find a doc value.

**Comment by dvirsky** [comment]
@tw-bert when we get to that stage we don't know what the index value was...

**Comment by tw-bert** [comment]
@dvirsky Ah okay, it was just a thought. Using `WITHSORTKEYS` you can already get the *sort* value when there is no doc, but apparently it's cumbersome for index values. I'll leave the better implementation ideas to you guys ;).

**Comment by dvirsky** [comment]
Oh, I forgot to mention that if the properties are sortable you can extract
them just by doing something like `APPLY @num AS num`!

**Comment by tw-bert** [comment]
@dvirsky Good info. I have to play around a bit using this for the scenario above, and check if making fields sortable for this reason alone would waste too much mem.

**Comment by paulflo150** [comment]
@dvirsky just to make sure I understand how to use FT.AGGREGATE vs FT.SEARCH. In order for FT.AGGREGATE to return the hash index key, all I would need to do is add the key as a Sortable field to my Schema, and use APPLY to return it? Although FT.SEARCH works in most of my cases, I do have some scenarios where I need to sort by multiple fields.

Thanks!

**Comment by paulflo150** [comment]
Trying to use FT.AGGREGATE instead of FT.SEARCH, but given the syntax below it does not matter what I am passing to SORTBY, the results appear to always be sorted by the first SORTABLE field found in the index (in this case ClientID).

`FT.AGGREGATE clients_hash_index_2 "*" SORTBY 4 @LastName ASC @FirstName ASC LIMIT 0 10 APPLY @ClientID as ClientID`

This is what my index looks like:
`FT.INFO clients_hash_index_2
 1) index_name
 2) clients_hash_index_2
 3) index_options
 4) (empty list or set)
 5) fields
 6)  1) 1) ClientID
        2) type
        3) NUMERIC
        4) SORTABLE
     2) 1) FirstName
        2) type
        3) TEXT
        4) WEIGHT
        5) "1"
        6) SORTABLE
     3) 1) LastName
        2) type
        3) TEXT
        4) WEIGHT
        5) "1"
        6) SORTABLE`

FT.SEARCH on the other hand works:
`FT.SEARCH clients_hash_index_2 "*" SORTBY @LastName ASC LIMIT 0 10`

Any ideas what's missing from the FT.AGGREGATE?

**Comment by itamarhaber** [comment]
@paulflo150 please open a new issue with the latest finding.

Heads up: I tried reproducing and found that if you do a dummy APPLY before SORTBY for the fields it should work. That said, this isn't a nice behavior and should be fixed /cc @mnunberg
